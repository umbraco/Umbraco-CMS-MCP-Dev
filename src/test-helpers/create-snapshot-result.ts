import { BLANK_UUID } from "@/constants/constants.js";
import { CallToolResult } from "@modelcontextprotocol/sdk/types.js";

export function createSnapshotResult(result: any, idToReplace?: string) {
  if (!result?.content) {
    return result;
  }

  function normalizeItem(i: any) {
    const item = { ...i, id: BLANK_UUID };
    if (item.parent) {
      item.parent = { ...item.parent, id: BLANK_UUID };
    }
    if (item.createDate) {
      item.createDate = "NORMALIZED_DATE";
    }
    if (item.publishDate) {
      item.publishDate = "NORMALIZED_DATE";
    }
    if (item.updateDate) {
      item.updateDate = "NORMALIZED_DATE";
    }
    return item;
  }

  return {
    ...result,
    content: result.content.map((item: any) => {
      if (item.type === "text") {
        if (idToReplace) {
          // For single item responses
          let text = item.text.replace(idToReplace, BLANK_UUID);
          try {
            const parsed = JSON.parse(text);
            if (parsed.createDate) {
              parsed.createDate = "NORMALIZED_DATE";
            }
            if (parsed.availableUntil) {
              parsed.availableUntil = "NORMALIZED_DATE";
            }
            if (parsed.publishDate) {
              parsed.publishDate = "NORMALIZED_DATE";
            }
            if (parsed.updateDate) {
              parsed.updateDate = "NORMALIZED_DATE";
            }
            if (parsed.variants && Array.isArray(parsed.variants)) {
              parsed.variants = parsed.variants.map((variant: any) => {
                if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
                if (variant.publishDate)
                  variant.publishDate = "NORMALIZED_DATE";
                if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
                return variant;
              });
            }
            text = JSON.stringify(parsed);
          } catch {}
          return {
            ...item,
            text,
          };
        } else {
          // For list responses
          const parsed = JSON.parse(item.text);
          if (Array.isArray(parsed)) {
            // Handle ancestors API response
            return {
              ...item,
              text: JSON.stringify(parsed.map(normalizeItem)),
            };
          }
          // Handle other list responses
          if (parsed.items) {
            parsed.items = parsed.items.map(normalizeItem);
          }
          if (parsed.variants && Array.isArray(parsed.variants)) {
            parsed.variants = parsed.variants.map((variant: any) => {
              if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
              if (variant.publishDate) variant.publishDate = "NORMALIZED_DATE";
              if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
              return variant;
            });
          }
          return {
            ...item,
            text: JSON.stringify(parsed),
          };
        }
      }
      return item;
    }),
  };
}

export function normalizeErrorResponse(result: CallToolResult): CallToolResult {
  if (
    Array.isArray(result.content) &&
    result.content[0]?.text &&
    typeof result.content[0].text === "string"
  ) {
    // Replace any traceId in the text with a normalized version
    result.content[0].text = result.content[0].text.replace(
      /00-[0-9a-f]{32}-[0-9a-f]{16}-00/g,
      "normalized-trace-id"
    );
  }
  return result;
}
