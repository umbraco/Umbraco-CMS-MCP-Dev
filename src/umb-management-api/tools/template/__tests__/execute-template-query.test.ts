import { postTemplateQueryExecuteBody } from "@/umb-management-api/umbracoManagementAPI.zod.js";
import ExecuteTemplateQueryTool from "../post/execute-template-query.js";
import { createSnapshotResult } from "@/test-helpers/create-snapshot-result.js";
import { jest } from "@jest/globals";

describe("execute-template-query", () => {
  let originalConsoleError: typeof console.error;

  beforeEach(() => {
    originalConsoleError = console.error;
    console.error = jest.fn();
  });

  afterEach(async () => {
    console.error = originalConsoleError;
  });

  it("should execute a simple template query", async () => {
    const queryBody = postTemplateQueryExecuteBody.parse({
      rootDocument: null,
      documentTypeAlias: null,
      filters: [],
      sort: null,
      take: 10
    });

    const result = await ExecuteTemplateQueryTool().handler(queryBody, {
      signal: new AbortController().signal,
    });

    expect(createSnapshotResult(result)).toMatchSnapshot();
  });

  it("should execute a template query with document type filter", async () => {
    const queryBody = postTemplateQueryExecuteBody.parse({
      rootDocument: null,
      documentTypeAlias: "contentPage",
      filters: [],
      sort: null,
      take: 5
    });

    const result = await ExecuteTemplateQueryTool().handler(queryBody, {
      signal: new AbortController().signal,
    });

    expect(createSnapshotResult(result)).toMatchSnapshot();
  });

  it("should execute a template query with filters and sorting", async () => {
    const queryBody = postTemplateQueryExecuteBody.parse({
      rootDocument: null,
      documentTypeAlias: null,
      filters: [
        {
          propertyAlias: "umbracoNaviHide",
          operator: "NotEquals",
          constraintValue: "1"
        }
      ],
      sort: {
        propertyAlias: "createDate",
        direction: "Descending"
      },
      take: 10
    });

    const result = await ExecuteTemplateQueryTool().handler(queryBody, {
      signal: new AbortController().signal,
    });

    expect(createSnapshotResult(result)).toMatchSnapshot();
  });

  it("should handle invalid query gracefully", async () => {
    const queryBody = postTemplateQueryExecuteBody.parse({
      rootDocument: null,
      documentTypeAlias: "nonExistentType",
      filters: [],
      sort: null,
      take: 10
    });

    const result = await ExecuteTemplateQueryTool().handler(queryBody, {
      signal: new AbortController().signal,
    });

    expect(createSnapshotResult(result)).toMatchSnapshot();
  });
});