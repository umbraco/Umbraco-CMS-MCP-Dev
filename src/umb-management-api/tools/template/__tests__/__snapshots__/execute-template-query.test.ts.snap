// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute-template-query should execute a simple template query 1`] = `
{
  "content": [
    {
      "text": "{"queryExpression":"Umbraco.ContentAtRoot().FirstOrDefault()\\n    .Children()\\n    .Where(x => x.IsVisible())\\n    .Take(10)","sampleResults":[{"icon":"icon-document color-blue","name":"Features"},{"icon":"icon-document color-blue","name":"About"},{"icon":"icon-thumbnail-list color-blue","name":"Blog"},{"icon":"icon-message color-blue","name":"Contact"},{"icon":"icon-search color-blue","name":"Search"},{"icon":"icon-users color-blue","name":"Authors"}],"resultCount":6,"executionTime":0}",
      "type": "text",
    },
  ],
}
`;

exports[`execute-template-query should execute a template query with document type filter 1`] = `
{
  "content": [
    {
      "text": "{"queryExpression":"Umbraco.ContentAtRoot().FirstOrDefault()\\n    .ChildrenOfType(\\"contentPage\\")\\n    .Where(x => x.IsVisible())\\n    .Take(5)","sampleResults":[],"resultCount":0,"executionTime":0}",
      "type": "text",
    },
  ],
}
`;

exports[`execute-template-query should execute a template query with filters and sorting 1`] = `
{
  "content": [
    {
      "text": "{"queryExpression":"Umbraco.ContentAtRoot().FirstOrDefault()\\n    .Children()\\n    .Where(x => x.IsVisible())\\n    .OrderByDescending(x => x.createDate)\\n    .Take(10)","sampleResults":[{"icon":"icon-search color-blue","name":"Search"},{"icon":"icon-document color-blue","name":"Features"},{"icon":"icon-message color-blue","name":"Contact"},{"icon":"icon-thumbnail-list color-blue","name":"Blog"},{"icon":"icon-users color-blue","name":"Authors"},{"icon":"icon-document color-blue","name":"About"}],"resultCount":6,"executionTime":0}",
      "type": "text",
    },
  ],
}
`;

exports[`execute-template-query should handle invalid query gracefully 1`] = `
{
  "content": [
    {
      "text": "{"queryExpression":"Umbraco.ContentAtRoot().FirstOrDefault()\\n    .ChildrenOfType(\\"nonExistentType\\")\\n    .Where(x => x.IsVisible())\\n    .Take(10)","sampleResults":[],"resultCount":0,"executionTime":0}",
      "type": "text",
    },
  ],
}
`;
